### ISSUE TITLE
Displaying a Bitmap Image

### ISSUE THREAD

**Issue by nbulischeck** [psychon/x11rb#480]
I'm trying to display a raw bitmap image (ARGB8888, but I'm practicing with a small XBM from GIMP) using x11rb. I've read through the entire tutorial as well as a lot of X11 docs online and I'm still not quite sure how to do it (with x11rb).

Examples online use functions such as `xcb_image_create_native` and `xcb_create_pixmap_from_bitmap_data` or `xcb_image_create_from_bitmap_data` but those don't seem to be supported here.

From what I've seen, key aspects are using `put_image` as well as updating the window on every expose event using `copy_area`.

**Comment by psychon** [comment]
> Examples online use functions such as xcb_image_create_native and xcb_create_pixmap_from_bitmap_data or xcb_image_create_from_bitmap_data but those don't seem to be supported here.

All of these functions do not come from (lib)xcb, but from xcb-image. #316 tracks the addition of "something similar" to x11rb. X11 image formats is something that I do not understand well myself. I prefer letting cairo figure out those details.

> I'm trying to display a raw bitmap image (ARGB8888,

How important is the `A` for you? X11 by itself does not have any support for "anything with alpha". The Render extension adds operations with an alpha channel, but that means these are only available when the Render extension is available.

Since you seem to already have googled around a bit: Do you perhaps have a small, sample C program that I could port to x11rb for you? Anything like this?

**Comment by nbulischeck** [comment]
"Best" examples I've seen so far have been these two but as you mentioned they use functions from xcb-image :( I can live without the alpha channel.

https://github.com/enn/xcb-examples/blob/master/window_event_bitmap.c
https://stackoverflow.com/questions/43029489/how-to-display-an-image-into-an-xcb-window

**Comment by psychon** [comment]
Here is a patch to the `simple_window` example that makes it display some colors (red channel is "always fully on", green and blue increase in x / y direction). Perhaps this is already a good starting point.

![foo](https://user-images.githubusercontent.com/89482/84587922-924f9980-ae23-11ea-8b69-0f01499acd24.png)

```diff
diff --git a/examples/simple_window.rs b/examples/simple_window.rs
index 2646330..624c817 100644
--- a/examples/simple_window.rs
+++ b/examples/simple_window.rs
@@ -6,6 +6,30 @@ use x11rb::protocol::xproto::*;
 use x11rb::protocol::Event;
 use x11rb::wrapper::ConnectionExt as _;
 
+const IMG_WIDTH: u8 = 100;
+const IMG_HEIGHT: u8 = 100;
+
+// Interpolate from `from` to `to`. `value` is a value between `0` and `max` describing "how far
+// along" we are.
+fn interpolate(from: u8, to: u8, value: u8, max: u8) -> u8 {
+    let range = to as f32 - from as f32;
+    from + (0.5 + range * value as f32 / max as f32) as u8
+}
+
+fn make_image_data() -> Vec<u8> {
+    let mut result = Vec::new();
+    for y in 0..IMG_HEIGHT {
+        for x in 0..IMG_WIDTH {
+            let red = 0xff;
+            let green = interpolate(0, 0xff, y, IMG_HEIGHT);
+            let blue = interpolate(0, 0xff, x, IMG_WIDTH);
+            // I bet the order here depends on the system endianess...
+            result.extend(&[blue, green, red, 0]);
+        }
+    }
+    result
+}
+
 fn main() {
     let (conn, screen_num) = x11rb::connect(None).unwrap();
 
@@ -120,6 +144,14 @@ fn main() {
         match event {
             Event::Expose(event) => {
                 if event.count == 0 {
+                    // This does some assumptions about the supported pixmap formats that ever X11
+                    // server these days should support, but I bet "back in the day" there were
+                    // servers that would complain. Oh and this assumes the "standard" depth=24
+                    // visual is used and no idea if this works correctly on big endian.
+                    //
+                    // For extra efficiency, one could upload the image data just once to a pixmap
+                    // and then copy_area() from the pixmap to the window. I'm lazy instead.
+                    conn.put_image(ImageFormat::ZPixmap, win_id, gc_id, IMG_WIDTH.into(), IMG_HEIGHT.into(), 10, 10, 0, 24, &make_image_data()).unwrap();
                     // There already is a white background because we set background_pixel to white
                     // when creating the window.
                     let (width, height): (i16, i16) = (width as _, height as _);
```

**Comment by psychon** [comment]
> https://github.com/enn/xcb-examples/blob/master/window_event_bitmap.c

Ah, this one helps me. So, `xcb_create_pixmap_from_bitmap_data` is used (which only supports bitmaps, aka "you only have two colors"). The image is converted into what the X11 server can understand via `xcb_image_native` which checks if a conversion is necessary or not (that tells us which parts of the image format are important to compare) and then calls `xcb_image_convert` to do the actual conversion. This has a byte swapping special case and a "just use lots of 'get one pixel, write one pixel'"-loop, which will be quite slow.

Thanks for making me find these parts. I'll mention them in #316 (but I am still not sure if it makes sense to have such a generic utility library which is unnecessary in 99.9% of time but still does not just "always work" (urgh, color palettes)).

> https://stackoverflow.com/questions/43029489/how-to-display-an-image-into-an-xcb-window

Google gave me this one as well. :-)

The code in the question goes the "definitely correct" route, but also the "awfully slow" route. `xcb_image_put_pixel` does all the complicated dances necessary to work "more or less everywhere", but doing that dance per pixel means lots of code is executed per pixel.

The code in the question also does not really make sense to me. It uploads the image (`xcb_image_put`) before filling it with data? Or not? I am confused.

The code in the answer is "as bad" as my code above. It assumes that the pixel data can just be in BGRX order. Technically, X11 is from a time where everything was complicated. A visual describes how "pixel values" are interpreted as colors. These times, everything will do something like the BGRX that is assumed here, but X11 e.g. also supports color palettes and lots of other fun.

**Comment by psychon** [comment]
Did my reply help you or are you still lost?

**Comment by nbulischeck** [comment]
Yeah that definitely helped! My question now is: Is the limit for a put_image request capped at 255x255? I tried bumping it to 256 and turned the width + height into u16s, but I get a `Error(Length(RequestError...`.

**Comment by psychon** [comment]
The X11 server has a maximum request length. `Connection::maximum_request_bytes()` returns this number. In a terminal, `xdpyinfo | grep -i maximum` gives you the number as well. For me, that is `maximum request size:  16777212 bytes`. This should be enough for at least an image size of `1024x1024`.

Anyway, it just occurred that exceeding the maximum request size would result in a different error. You are getting a `BadLength` error from the X11 server. https://www.x.org/releases/X11R7.6/doc/xproto/x11protocol.html#id2623526 says that bad length means:

> The length of a request is shorter or longer than that required to minimally contain the arguments. The length of a request exceeds the maximum length accepted by the server.

(Ignore the second sentence, x11rb would catch that error locally before sending the request.)

So... I guess the values `width`, `height` and the length of the provided pixel data do not match. Hm... I can reproduce the problem by adding the following ontop of my patch above. I will investigate.
```diff
diff --git a/examples/simple_window.rs b/examples/simple_window.rs
index 624c817..1a8df0b 100644
--- a/examples/simple_window.rs
+++ b/examples/simple_window.rs
@@ -6,12 +6,12 @@ use x11rb::protocol::xproto::*;
 use x11rb::protocol::Event;
 use x11rb::wrapper::ConnectionExt as _;
 
-const IMG_WIDTH: u8 = 100;
-const IMG_HEIGHT: u8 = 100;
+const IMG_WIDTH: u16 = 256;
+const IMG_HEIGHT: u16 = 256;
 
 // Interpolate from `from` to `to`. `value` is a value between `0` and `max` describing "how far
 // along" we are.
-fn interpolate(from: u8, to: u8, value: u8, max: u8) -> u8 {
+fn interpolate(from: u8, to: u8, value: u16, max: u16) -> u8 {
     let range = to as f32 - from as f32;
     from + (0.5 + range * value as f32 / max as f32) as u8
 }
```
```
Error(Length(RequestError { error_code: 16, sequence: 18, bad_value: 33554433, minor_opcode: 0, major_opcode: 72 })))
Got an unexpected error
```

**Comment by nbulischeck** [comment]
Yep that's exactly what I'm getting. I also have the same maximum request size.

**Comment by psychon** [comment]
Huh. You found a bug in x11rb. I am not yet sure what exactly is going on, but something is wrong in the land of BigRequests.

Traditionally, X11 has a 16 bit length field that counts the number of 4-byte-quantities of the request. Thus, the maximum request length in "traditional X11" is 2^18 aka 256 KiB. A PutImage request for an area of 256x256 has 256*256*4+24 bytes, which is 2^18+24. To send such large requests, the BigRequests extension "brings some magic" so that the length field now has 32 bits.

I am sure that something around this is wrong, because after this first BadLength error, I get lots of follow-up errors that basically say stuff like "there is no request with number 225", so I guess this now interprets pixel data as a new request. Also, `xtrace` says that the `PutImage` request has length 262168 which... uhm... is correct (it is 2^18 + 24)... well...
```
sendmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="H\2\0\0\6\0\1\0", iov_len=8}, {iov_base="\0\0\300\2\1\0\300\2\200\0\0\2\n\0\n\0\0\30\0\0", iov_len=20}, {iov_base="\0\0\377\0\2\0\377\0\4\0\377\0\6\0\377\0\10\0\377\0\n\0\377\0\f\0\377\0\16\0\377\0"..., iov_len=262144}, {iov_base="", iov_len=0}], msg_iovlen=4, msg_controllen=0, msg_flags=0}, 0) = 219264
poll([{fd=3, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=3, revents=POLLOUT}])
sendmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="2\325\377\0004\325\377\0006\325\377\08\325\377\0:\325\377\0<\325\377\0>\325\377\0@\325\377\0"..., iov_len=42908}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 42908
munmap(0x7f711edc7000, 266240) 
```
The part beginning with `H\2\0\0` is the `PutImage` request. Let's see... the `u16` length field has value 0, which is correct, and the `u32` length field has value `\6\0\1\0` which is `2^16+6`, which is `2^18+24` divided by 4, thus also correct.

Okay, next, the writes: The first one writes 219264 bytes and the second one writes 42908 bytes. That's also the expected `2^18+24+4` bytes, but... I guess the length field should include the `+4` for the extra `u32` length field. Thus, the length field is off by one. Whoops.

Now that I have found the bug: The sad news is that you cannot work around this in your code (#483 will fix it). Thus, with the latest release of x11rb, you cannot send requests larger than `2^18` bytes. Thus, the image data in a PutImage request can be at most `2^18-24` bytes.

Here is an example on how to split up larger PutImage requests into smaller ones: https://gitlab.freedesktop.org/cairo/cairo/-/blob/aee96d175d8349945a6d1948a56abd4b4ec6ad84/src/cairo-xcb-connection-core.c#L120-160
(What this code does: It computes `(maximum request length - 24 - 4) / (4 * width)`. This is `rows`, the number of rows that can be sent in a single PutImage request. Then, the large `PutImage` request is split into multiple smaller ones.)

Thanks for finding this bug and sorry for finding this bug.

**Comment by nbulischeck** [comment]
I'm glad you were able to figure it out! And thanks for linking the example of multiple PutImage requests.

**Comment by psychon** [comment]
By the way, sorry for getting this issue closed. However, I feel like the original question was answered and the bug that was then found is now fixed (hm... so #483 does not really "fix" this issue, since this issue is not really about BigRequest being broken, but.... whatever).
Anyway, feel free to keep asking questions. I cannot promise anything, but I'll try to help. (The magic way to get me to disappear is to ask questions about keyboard input, because X11 and XKB are complicated. ;-) )

**Comment by psychon** [comment]
FYI: We just merged something like xcb's util-image in #507 (it can actually do more: It also tries to convert between different bit depths (but without any dithering or stuff like that), but that only works for non-grayscale and non-colormapped visuals; hopefully no one runs one of those these days).
